circom-types/src/groth16/zkey.rs

```diff
@@ -195,98 +215,74 @@ where
             b,
             c: vec![],
         };
+        // this thread automatically joins on the rayon scope, therefore we can
+        // only be here if the scope finished.
+        let vk = VerifyingKey {
+            alpha_g1: header.alpha_g1,
+            beta_g2: header.beta_g2,
+            gamma_g2: header.gamma_g2,
+            delta_g2: header.delta_g2,
+            // unwrap is fine, because we are guaranteed to have a Some value (rayon scope)
+            gamma_abc_g1: ic.unwrap()?,
+        };
 
-        Ok(ZKey { pk, matrices })
-    }
-}
-
-#[derive(Default, Clone, Debug)]
-pub(crate) struct ZVerifyingKey<P: Pairing> {
-    alpha_g1: P::G1Affine,
-    beta_g1: P::G1Affine,
-    beta_g2: P::G2Affine,
-    gamma_g2: P::G2Affine,
-    delta_g1: P::G1Affine,
-    delta_g2: P::G2Affine,
-}
-
-impl<P: Pairing + CircomArkworksPairingBridge> ZVerifyingKey<P>
-where
-    P::BaseField: CircomArkworksPrimeFieldBridge,
-    P::ScalarField: CircomArkworksPrimeFieldBridge,
-{
-    fn new<R: Read>(mut reader: R) -> ZKeyParserResult<Self> {
-        let alpha_g1 = P::g1_from_reader(&mut reader)?;
-        let beta_g1 = P::g1_from_reader(&mut reader)?;
-        let beta_g2 = P::g2_from_reader(&mut reader)?;
-        let gamma_g2 = P::g2_from_reader(&mut reader)?;
-        let delta_g1 = P::g1_from_reader(&mut reader)?;
-        let delta_g2 = P::g2_from_reader(&mut reader)?;
-
-        Ok(Self {
-            alpha_g1,
-            beta_g1,
-            beta_g2,
-            gamma_g2,
-            delta_g1,
-            delta_g2,
+        Ok(ZKey {
+            n_public: header.n_public,
+            beta_g1: header.beta_g1,
+            delta_g1: header.delta_g1,
+            // unwrap is fine, because we are guaranteed to have a Some value (rayon scope)
+            a_query: a_query.unwrap()?,
+            b_g1_query: b_g1_query.unwrap()?,
+            b_g2_query: b_g2_query.unwrap()?,
+            h_query: h_query.unwrap()?,
+            l_query: l_query.unwrap()?,
+            matrices,
+            vk,
         })
     }
 }
 
-#[derive(Clone, Debug)]
-struct HeaderGroth<P: Pairing> {
-    #[allow(dead_code)]
-    n8q: u32,
-    #[allow(dead_code)]
-    n8r: u32,
-
-    n_vars: usize,
-    n_public: usize,
-
-    domain_size: u32,
-    #[allow(dead_code)]
-    power: u32,
-
-    verifying_key: ZVerifyingKey<P>,
-}
-
 impl<P: Pairing + CircomArkworksPairingBridge> HeaderGroth<P>
 where
     P::BaseField: CircomArkworksPrimeFieldBridge,
     P::ScalarField: CircomArkworksPrimeFieldBridge,
 {
     fn read<R: Read>(mut reader: &mut R) -> ZKeyParserResult<Self> {
-        // TODO: Impl From<u32> in Arkworks
-        let n8q: u32 = u32::deserialize_uncompressed(&mut reader)?;
+        let _n8q: u32 = u32::deserialize_uncompressed(&mut reader)?;
```

Comment:
Should we assert/err on the expected value of these?

Reply:
I thought about that but I if everything else is correct it doesnt matter what those values are. Therefore I just ignored them. Maybe put a warning log in there?
We maybe want logging anyways?

Reply:
We should probably add a lot more tracing in general in a new pr

Reply:
They still have some expected value though and if they are not that then there is some fuckery going on and we should probably error